Article about functional programming in kotlin (#133)
Zuehlke/software-engineering-corner: src/articles/kotlin-functional-programming/index.md

released: '2025-11-12T16:20:00.243Z'
cover: images/fp.png
author: Patrick Wilmes
tags:
  - functional programming
  - Kotlin
  - General Programming
  - Software Engineering
shortDescription: >-
    An article about utilizing the strength of Object Oriented Programming with
    Functional Programming. Combining the best out of both worlds.
---

> **"Every tool is a lens. The more lenses you collect, the more truth you see."**

Most experienced developers live and breathe **object-oriented programming (OOP)**.
We design systems in terms of entities, behaviors, and hierarchies; we reason about state, lifecycle, and encapsulation. It works beautifully—until it doesn’t.

At some point, your code starts fighting back. Mutable state causes race conditions.
Business logic sprawls across layers. Unit tests turn into integration monsters.
You find yourself wishing you could express _what_ you want, without micromanaging _how_ it happens.

That’s where **functional programming (FP)** enters — not as a rival philosophy, but as a complementary one. FP isn’t about writing math puzzles or memorizing category theory.

> It’s a practical way to make logic predictable, composable, and testable.

Reviewer: ```suggestion
It’s a practical way to predict, compose and test business logic.
```

Instead of modeling _objects that do things_, FP focuses on _functions that transform data_.

> This article helps experienced OOP developers understand and apply FP concepts using **Kotlin**, without detouring into abstract theory. But the same principles and techniques can

Reviewer: ```suggestion
This article helps experienced OOP developers understand and apply FP concepts using **Kotlin**, without detouring into abstract theory.
```

- The logic is deterministic.

> For this to stay pure, all nested properties in `DepartmentFee` must also be immutable.

### Why Purity Matters

Pure functions are easier to:
- **Test** – no mocks, no hidden dependencies
- **Reason about** – behavior is explicit
- **Parallelize** – no shared state
- **Cache** – same input → same output
- **Compose** – composing pure functions yields pure results

Purity scales reasoning: you can trust each function in isolation.

---

## Immutability

Functional programming thrives on **immutability** — once a value is created, it never changes.  
This removes entire classes of bugs caused by hidden mutations.

```kotlin
val hello = "value"
hello = "other value" // ❌ Compile error
```

Kotlin’s `val` is your ally. Prefer it whenever possible.

> Immutability removes “time” as a hidden parameter—your data means the same thing forever.

### When Is Mutability Acceptable?

Sometimes mutability is fine—if it’s **encapsulated** and invisible from the outside.

Example: optimizing a collection transformation.

```kotlin
fun <T, R> Collection<T>.mapToSet(mappingFunc: (T) -> R): Set<R> {
    val targetSet = mutableSetOf<R>()

>     forEach { targetSet.add(mappingFunc(it)) }

Reviewer: maybe shortly explain that the scope changes in extension function to the type which we extend ;)

    if (b == 0) throw IllegalArgumentException("Division by zero")
    return a / b
}
```

This is impure — it throws exceptions.

### The Functional Alternative: `Result` Type

```kotlin
sealed class Result<out T> {
    data class Success<T>(val value: T) : Result<T>()
    data class Failure(val error: Throwable) : Result<Nothing>()
}

fun divide(a: Int, b: Int): Result<Int> =
    if (b == 0)
        Result.Failure(IllegalArgumentException("Division by zero"))
    else
        Result.Success(a / b)

when (val result = divide(10, 0)) {
    is Result.Success -> println("Result: ${result.value}")
    is Result.Failure -> println("Error: ${result.error.message}")
}
```

Now the function is pure — no hidden control flow, no exceptions.

---

## Monads Without Fear

You can think of a **monad** as a _computation context_ — a box that carries both a value and its meaning  
(was it successful, optional, deferred?).

A monad provides three operations:
1. A way to create the context (`of`, `pure`)
2. `map` to transform the inner value
3. `flatMap` to chain operations that return new contexts

Let’s build a simple one.

```kotlin
data class Box<T>(val value: T) {
    companion object {

>         fun <T> of(value: T): Box<T> = Box(value)

Reviewer: additionally I would argue, that this function is overhead and does not add value.

}
```

**Monads let you:**
- Chain operations safely
- Avoid nested conditionals or `try/catch` pyramids
- Keep business logic clean
- Maintain purity—no hidden effects

Common monads you’ll encounter:
- `Option` / `Maybe` – presence or absence
- `Result` / `Either` – success or failure
- `List` – non-deterministic results
- `IO` – deferred side effects

---

## When to Use Which Paradigm

### Functional Programming Shines When
- You’re transforming data
- Logic is complex but composable
- You need thread safety
- You want testability and predictability

### OOP Shines When
- Modeling domain entities (User, Order, etc.)
- Managing lifecycles or state
- Integrating with imperative APIs (Android, JavaFX, etc.)

### Combine Both

Use classes for structure, functions for logic, and monads for effects.

```kotlin
// OOP: Structure
data class Order(val items: List<OrderItem>, val customerId: String)
data class OrderItem(val productId: String, val quantity: Int, val price: Double)

// FP: Logic
fun calculateTotal(order: Order): Double =
    order.items.map { it.quantity * it.price }.sum()

fun applyDiscount(order: Order, discount: Double): Order =

>     order.copy(items = order.items.map { it.copy(price = it.price * (1 - discount)) }))

Reviewer: ```suggestion
    this.copy(items = order.items.map { it.copy(price = it.price * (1 - discount)) }))
```

// Combine Both
class OrderService {
    fun processOrder(order: Order): Result<Order> =
        validateOrder(order)
            .map { applyDiscount(it, 0.1) }

>             .flatMap { saveOrder(it) } // Side effect wrapped in Result

Reviewer: ```suggestion
            .flatMap { save() } // Side effect wrapped in Result
```

}
```

A backend service might model its domain in OOP but express its data transformations functionally—clean logic, explicit effects.

---

## Conclusion

Functional programming doesn’t replace object-oriented design—it **completes** it.  
When OOP gives us _nouns_, FP gives us _verbs_. Together they form the full grammar of software.

The goal isn’t _purity_, but _clarity_.  
Start small: use `val` instead of `var`, extract pure functions, embrace `map`, `filter`, and `flatMap`.  

> Over time, you’ll find that FP principles don’t just make your code safer—they make it _simpler_.

Reviewer: ```suggestion
Over time, you’ll find that FP principles don’t just make your code safer—they also make it _simpler_.
```

lifecycle, and encapsulation. It works beautifully—until it doesn’t.

At some point, your code starts fighting back. Mutable state causes race conditions.
Business logic sprawls across layers. Unit tests turn into integration monsters.
You find yourself wishing you could express _what_ you want, without micromanaging _how_ it happens.

That’s where **functional programming (FP)** enters — not as a rival philosophy, but as a complementary one. FP isn’t about writing math puzzles or memorizing category theory.
It’s a practical way to make business logic predictable, composable, and testable.
Instead of modeling _objects that do things_, FP focuses on _functions that transform data_.

This article helps experienced OOP developers understand and apply FP concepts using **Kotlin**, without detouring into abstract theory.
The same principles and techniques can also be applied to any Java 8+ code base.
You’ll learn how to combine both paradigms in one codebase to build systems that are easier to reason about, safer to change, and more expressive.

---

## Functional and Object-Oriented: Not Opposites, but Complements

> There’s no single “superior” paradigm. Paradigms are tools, each emphasizing a different aspect of design.

Reviewer 2: ```suggestion
There’s no single “superior” paradigm. Paradigms are tools, each emphasizing a different aspect of design.

```

- **OOP** models the world with _entities_ and _behavior_. It gives structure and relationships.    
- **FP** models _data transformations_. It focuses on purity, predictability, and composition.

> Viewed together:

Reviewer 2: ```suggestion
Viewed together:

```

> > **Use OOP to give your project structure, and FP to express logic clearly.**

Reviewer 2: ```suggestion
> **Use OOP to give your project structure, and FP to express logic clearly.**

```

The combination yields systems that are both architecturally coherent and mathematically predictable — code that’s easier to test, reason about, and refactor.

---

## Pure Functions

A **pure function** always returns the same output for the same input and produces no side effects.  
This property — **referential transparency** — means you can replace a function call with its result without changing the program’s behavior.

### What Makes a Function Pure?

> A pure function must satisfy two requirements:

Reviewer 2: ```suggestion
A pure function must satisfy two requirements:

```

1. **Deterministic** – Output depends only on inputs.
2. **No side effects** – The function doesn’t:
    - Modify global or static state
    - Mutate inputs or external objects
    - Perform I/O (network, database, file I/O, logging)
    - Throw exceptions or rely on them for control flow
    - Access non-deterministic data (time, random numbers)

```kotlin
fun addTax(price: Double, taxRate: Double): Double {
    return price * (1 + taxRate)
}

// Always returns the same result for the same inputs
addTax(100.0, 0.19) // 119.0
addTax(100.0, 0.19) // 119.0
```

A more realistic example:

```kotlin
fun setMandateIfBlank(
    departmentFee: DepartmentFee,
    identifier: String
): DepartmentFee {
    return if (departmentFee.irmaMandate.isNullOrBlank()) {
        departmentFee.copy(irmaMandate = identifier)
    } else {
        departmentFee
    }
}
```

> This is pure because:

Reviewer 2: ```suggestion
This is pure because:

```

- `DepartmentFee` (a `data class`) is immutable.
- We don’t mutate it; we use `copy()`.
- The logic is deterministic.

> For this to stay pure, all nested properties in `DepartmentFee` must also be immutable.

### Why Purity Matters

> Pure functions are easier to:

Reviewer 2: ```suggestion
Pure functions are easier to:

```

- **Test** – no mocks, no hidden dependencies
- **Reason about** – behavior is explicit
- **Parallelize** – no shared state
- **Cache** – same input → same output
- **Compose** – composing pure functions yields pure results

Purity scales reasoning: you can trust each function in isolation.

---

## Immutability

Functional programming thrives on **immutability** — once a value is created, it never changes.  
This removes entire classes of bugs caused by hidden mutations.

```kotlin
val hello = "value"
hello = "other value" // ❌ Compile error
```

Kotlin’s `val` is your ally. Prefer it whenever possible.

> Immutability removes “time” as a hidden parameter—your data means the same thing forever.

### When Is Mutability Acceptable?

Sometimes mutability is fine—if it’s **encapsulated** and invisible from the outside.

Example: optimizing a collection transformation.

```kotlin
fun <T, R> Collection<T>.mapToSet(mappingFunc: (T) -> R): Set<R> {
    val targetSet = mutableSetOf<R>()
    this.forEach { targetSet.add(mappingFunc(it)) }
    return targetSet
}
```

This function:

> - Uses internal mutability for performance

Reviewer 2: ```suggestion

- Uses internal mutability for performance
```

val doubled = multiplyBy2(5) // 10
```

A **higher-order function** takes other functions as arguments, or returns them.

```kotlin
val list = listOf(1, 2, 3)
list.map { it + 1 } // Higher-order function
```

Defining one yourself:

```kotlin
fun createTransformer(
    mapper: (Int) -> String,
    multiplier: Int
): (Int) -> String {
    return { input -> mapper(input * multiplier) }
}

val transformer = createTransformer({ it.toString() }, 10)
transformer(5) // "50"
```

### Functions as Parameters

```kotlin
fun isEven(number: Int): Boolean = number % 2 == 0
val numbers = listOf(1, 2, 3, 4, 5)
numbers.filter(::isEven) // [2, 4]
```

The `::` operator passes function references directly.

### Returning Functions (Lazy Evaluation)

Returning functions lets you **defer computation**.

```kotlin
import kotlin.random.Random

fun determineSum(a: Int, b: Int): () -> Int {
    println("Function created, but not calculated yet")
    return {
        println("Now calculating...")
        a + b
    }
}

val shouldCalculate = Random.nextBoolean()
val calculation = determineSum(10, 10)

if (shouldCalculate) {
    val result = calculation()
    println(result)
}
```

> **Benefits:**

Reviewer 2: ```suggestion
**Benefits:**

```

    val loadPaymentHistory: () -> List<Payment> = { database.loadPaymentHistory(userId) }

    return user.copy(
        profileLoader = loadFullProfile,
        paymentLoader = loadPaymentHistory
    )
}
```

---

## Function Composition

In OOP, we often compose behavior through inheritance or strategy patterns.
In FP, we use **function composition** — building complex logic by chaining smaller functions.

```kotlin
fun <A, B, C> compose(f: (B) -> C, g: (A) -> B): (A) -> C {
    return { a -> f(g(a)) }
}

val addOne: (Int) -> Int = { it + 1 }
val multiplyBy2: (Int) -> Int = { it * 2 }

val addOneThenDouble = compose(multiplyBy2, addOne)
addOneThenDouble(5) // (5 + 1) * 2 = 12
```

Composition replaces many classical patterns (Decorator, Strategy) with simpler, type-safe building blocks.

---

## Handling Side Effects and Errors: From Exceptions to Monads

In an imperative world, exceptions are invisible branches in your control flow.  
FP replaces invisibility with _explicitness_: errors are just data.

```kotlin
fun divide(a: Int, b: Int): Int {
    if (b == 0) throw IllegalArgumentException("Division by zero")
    return a / b
}
```

> This is impure — it throws exceptions.

Reviewer 2: ```suggestion
This is impure — it throws an exception.
```

### The Functional Alternative: `Result` Type

```kotlin
sealed class Result<out T> {
    data class Success<T>(val value: T) : Result<T>()
    data class Failure(val error: Throwable) : Result<Nothing>()
}

fun divide(a: Int, b: Int): Result<Int> =
    if (b == 0)
        Result.Failure(IllegalArgumentException("Division by zero"))
    else
        Result.Success(a / b)

when (val result = divide(10, 0)) {
    is Result.Success -> println("Result: ${result.value}")
    is Result.Failure -> println("Error: ${result.error.message}")
}
```

Now the function is pure — no hidden control flow, no exceptions.

---

## Monads Without Fear

You can think of a **monad** as a _computation context_ — a box that carries both a value and its meaning  
(was it successful, optional, deferred?).

> A monad provides three operations:

Reviewer 2: ```suggestion
A monad provides three operations:

```

    data class Success<T>(val value: T) : Result<T>()
    data class Failure(val error: Throwable) : Result<Nothing>()

    fun <R> map(transform: (T) -> R): Result<R> =
        when (this) {
            is Success -> Success(transform(value))
            is Failure -> this
        }

    fun <R> flatMap(transform: (T) -> Result<R>): Result<R> =
        when (this) {
            is Success -> transform(value)
            is Failure -> this
        }
}
```

Now we can chain cleanly.

```kotlin
fun parseNumber(input: String): Result<Int> =
    try {
        Result.Success(input.toInt())
    } catch (e: NumberFormatException) {
        Result.Failure(e)
    }

fun safeDivide(a: Int, b: Int): Result<Int> =
    if (b == 0)
        Result.Failure(IllegalArgumentException("Division by zero"))
    else
        Result.Success(a / b)

val result = parseNumber("20")
    .flatMap { numerator ->
        parseNumber("4").flatMap { denominator ->
            safeDivide(numerator, denominator)
        }
    }
    .map { it * 2 }

when (result) {
    is Result.Success -> println("Final result: ${result.value}")
    is Result.Failure -> println("Error: ${result.error.message}")
}
```

> **Monads let you:**

Reviewer 2: ```suggestion
**Monads let you:**

```

- Chain operations safely
- Avoid nested conditionals or `try/catch` pyramids
- Keep business logic clean
- Maintain purity—no hidden effects

> Common monads you’ll encounter:

Reviewer 2: ```suggestion
Common monads you’ll encounter:

```

- `Option` / `Maybe` – presence or absence
- `Result` / `Either` – success or failure
- `List` – non-deterministic results
- `IO` – deferred side effects

---

## When to Use Which Paradigm

> ### Functional Programming Shines When

Reviewer 2: ```suggestion
### Functional Programming Shines When

```

Reviewer 2: Instead of using titles, I would have used a different format:

```
Functional programming shines when ...

- ... you're tranforming data.
- ... logic is complex but composable
- ...
```

- You’re transforming data
- Logic is complex but composable
- You need thread safety
- You want testability and predictability

> ### OOP Shines When

Reviewer 2: ```suggestion
### OOP Shines When

```
